service/graph.go

```diff
@@ -0,0 +1,80 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service // import "go.opentelemetry.io/collector/service"
+
+import (
+	"context"
+	"net/http"
+
+	"gonum.org/v1/gonum/graph/simple"
+	"gonum.org/v1/gonum/graph/topo"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+var _ pipelines = (*serviceGraph)(nil)
+
+type serviceGraph struct {
+
+	// All component instances represented as nodes, with directed edges indicating data flow.
+	componentGraph *simple.DirectedGraph
+}
+
+func (g *serviceGraph) StartAll(ctx context.Context, host component.Host) error {
+	nodes, err := topo.Sort(g.componentGraph)
```

Comment:
Should we do this in the "constructor" to avoid doing it during start and shutdown?

Reply:
We could but in my opinion I prefer this way because then we have a single source of truth, instead of maintaining a separate representation. We have a handle to the graph and a very easy way to sort it on demand. Hypothetically, this design is more resilient in case we later support dynamically updating the graph.